src/librustc_typeck/check/demand.rs

```diff
@@ -239,6 +239,19 @@ impl<'a, 'gcx, 'tcx> FnCtxt<'a, 'gcx, 'tcx> {
                 }
                 None
             }
+            (expected_ty, &ty::TyRef(_, checked_ty)) => {
+                if expected_ty == &checked_ty.ty.sty {
+                    let sp = self.sess().codemap().call_span_if_macro(expr.span);
+                    if let Ok(src) = self.tcx.sess.codemap().span_to_snippet(sp) {
+                        return if src.starts_with("&") {
+                            Some(format!("try with `{}`", &src[1..]))
+                        } else {
+                            Some(format!("try with `*{}`", &src))
```

Comment:
For this case, I'd need to check if it's actually possible to deref without error but I don't how to do it. Anyone has an idea?